### PR TITLE
Add separate send queue for kbfs file edit messages skipping the outbox

### PR DESCRIPTION
**Pull request by joshblum**

### CODE REVIEW

**Comment by strib** [comment]
Can you please wait for me to review this as well before you merge?  I'll look tomorrow.

**Comment by mmaxim** [comment]
Probably can't work this way, I'd like the calls from KBFS to still return to it immediately.

File: go/chat/deliverer.go [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+				ConvID: convID,
+			})
+			s.G().ActivityNotifier.Activity(ctx, uid, obr.Msg.ClientHeader.Conv.TopicType, &act,
+				chat1.ChatActivitySource_LOCAL)
+		}))
+	s.outbox.SetClock(s.clock)
+
+	s.delivering = true
+	s.shutdownCh = make(chan struct{})
+	s.eg.Go(func() error { return s.deliverLoop(s.shutdownCh) })
```
Comment (joshblum) [inline go/chat/deliverer.go:176]: changed to support multiple loops

File: go/chat/deliverer.go [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+	return s.doStop(ctx)
+}
+
+func (s *Deliverer) doStop(ctx context.Context) chan struct{} {
+	cb := make(chan struct{})
+	if s.delivering {
+		s.Debug(ctx, "stopping")
+		close(s.shutdownCh)
+		s.delivering = false
+		go func() {
```
Comment (joshblum) [inline go/chat/deliverer.go:192]: block cb until loops finish

File: go/chat/deliverer.go [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+	return s.delivering
+}
+
+func (s *Deliverer) Queue(ctx context.Context, convID chat1.ConversationID, msg chat1.MessagePlaintext,
+	outboxID *chat1.OutboxID, sendOpts *chat1.SenderSendOptions, prepareOpts *chat1.SenderPrepareOptions,
+	identifyBehavior keybase1.TLFIdentifyBehavior) (obr chat1.OutboxRecord, err error) {
+	defer s.Trace(ctx, func() error { return err }, "Queue")()
+
+	// KBFSFILEEDIT msgs skip the traditional outbox
+	if msg.ClientHeader.Conv.TopicType == chat1.TopicType_KBFSFILEEDIT {
```
Comment (joshblum) [inline go/chat/deliverer.go:254]: skip outbox for kbfs files

File: go/chat/deliverer.go [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+func (s *Deliverer) shouldBreakLoop(ctx context.Context, obr chat1.OutboxRecord) bool {
+	if obr.Msg.ClientHeader.MessageType == chat1.MessageType_UNFURL {
+		s.Debug(ctx, "shouldBreakLoop: not breaking deliverer loop for unfurl failure: outboxID: %s",
+			obr.OutboxID)
+		return false
+	}
+	return true
+}
+
+func (s *Deliverer) kbfsDeliverLoop(shutdownCh chan struct{}) error {
```
Comment (joshblum) [inline go/chat/deliverer.go:660]: read off of kbfs queue and do send

**Comment by strib** [comment]
With no more KBFS changes, there's no need to wait for my review anymore.  But I'm curious how this fixes the actual deadlock in the original report.  is it just an orthogonal fix that will slow down the processing of kbfs edit requests?  Or does it fix the actual, persistent deadlock observed in that issue?

**Comment by joshblum** [comment]
@strib I'll look further into the deadlock, this PR is just to address the load on the chat system.

**Pull request merged**